syntax = "proto3";

package helloworld;

import "google/api/annotations.proto";
option go_package = "helloworld/helloworld";

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply)  {
        option (google.api.http) = {
            get: "/helloworld/{name=test}",
            additional_bindings {
                post: "/say_hello",
                body: "*",
            }
        };
  }

  rpc SayMulti (MultiRequest) returns (HelloReply)  {
    option (google.api.http) = {
        get: "/helloworld/{inner.inner_name}",
        additional_bindings {
            post: "/say_multi",
            body: "*",
        }
    };
  }
  rpc Echo (MultiRequest) returns (HelloReply)  {
    option (google.api.http) = {
        get: "/echo/{inner.inner_name=echo/*}",
        additional_bindings {
            post: "/echo",
            body: "*",
        }
    };
  }
}

message Inner {
  string inner_name = 1;
}


message HelloRequest {
  string name = 1;
}


// The request message containing the user's name.
message MultiRequest {
  string name = 1;
  Inner inner = 2;
  repeated int32 nums = 3;
  map<string, string> metadata = 4;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}